### PR TITLE
Feature 1203 ioda2nc

### LINKED ISSUE #1101: Update wrapped MET config files to reference MET_TMP_DIR in tmp value
The default value is currently /tmp, however some machines discourage using this directory to write temp files. When running through METplus wrappers, the environment variable MET_TMP_DIR is set which overrides this value. Running with the wrapped MET config files outside of the wrappers (via command line) require the other environment variables to be set to run or it will fail. If all of the env vars set for a command are set, then it will include MET_TMP_DIR, however this may not always be the case. Setting the value for tmp to $(MET_TMP_DIR) will ensure that the value is overridden properly when running by hand and make it clear to users that the value set for tmp in those files is not being used.

## Describe the Enhancement ##
Change the files in parm/met_config to reference the MET_TMP_DIR environment variable instead of /tmp.

### Time Estimate ###
~< 1 day

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [X] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
4.1.0

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [X] Select **component(s)**
- [X] Select **priority**
- [X] Select **requestor(s)**

### Projects and Milestone ###
- [X] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [X] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [X] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### LINKED ISSUE #1203: New Wrapper: IODA2NC
*Replace italics below with details for this issue.*

## Describe the New Feature ##
The IODA2NC tools provides one step in the processing chain to "evaluate the model on it's native domain" using observations ingested by the JEDI DA system.  We now need to wrap it with METplus and then developed a use-case (https://github.com/dtcenter/METplus/issues/1204).

A description of the IODA2NC tool can be found here: https://met.readthedocs.io/en/main_v10.0/Users_Guide/reformat_point.html#ioda2nc-tool

Being there aren't too many settings, it would probably be good to include all command line options and configuration options (share/met/config/IODA2NcConfig_default)

### Acceptance Testing ###
Start with using the current tool test data. 
It might be good to also get some sample data from running JEDI on MPAS runs.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
needs to be in METplus v4.1.0

### Funding Source ###
2702691

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### LINKED ISSUE #1204: New Use Case: IODA2NC
*Replace italics below with details for this issue.*

## Describe the New Use Case ##
Once IODA2NC is wrapped (https://github.com/dtcenter/METplus/issues/1203), we need to develop a use-case to demonstrate how to use it.  

### Use Case Name and Category ###
*Provide use case name, following Contributor's Guide naming template, and list which category the use case will reside in.*
*If a new category is needed for this use case, provide its name and brief justification*

This should go in the METplus Tools section
IODA2NC/IODA2NC.conf

### Input Data ###
test data

### Acceptance Testing ###
*Describe tests required for new functionality.*
*As use case develops, provide a run time here*

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
needs to be in METplus v4.1.0

### Funding Source ###
2702691

### Assignee ###
- [x ] Select **engineer(s)** or **no engineer** required
- [x ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ x] Select **component(s)**
- [x ] Select **priority**
- [ ] Select **requestor(s)**
- [ ] Select **privacy**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Use Case Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept your changes. Merge branch into **develop**.
- [ ] Create a second pull request to merge **develop** into **develop-ref**, following the same steps for the first pull request.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### LINKED ISSUE #1212: Properly handle list values that include square braces
Multiple users have defined list values in METplus config files including the square braces that are found in the MET config files:

POINT_STAT_MASK_SID = ["/data/users/cfrd/METplus_playaround/StationLists/2011.stns", "/data/users/cfrd/METplus_playaround/StationLists/2012.stns"]

List values in METplus config files currently cannot not include square braces and will be formatted properly when set in the MET config files.

## Describe the Enhancement ##
Enhance the list parsing logic to properly handle list values that contain the square braces so that the above example will work properly.

### Time Estimate ###
<1 day

### Sub-Issues ###
None

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [X] Select **engineer(s)** or **no engineer** required
- [X] **no scientist** required

### Labels ###
- [X] Select **component(s)**
- [X] Select **priority**
- [X] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [X] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [X] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
This PR will close a few issues:
* #1203
* #1204 
* #1101 (Commit: 10f17e7)
* #1212 (Commit: 19393eb)

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

Added unit tests to ensure commands are built properly for multiple run times, optional command line arguments are added properly, and environment variables referenced in the wrapped MET config files are set properly.
Added a new use case example IODA2NC.conf based on a MET unit test

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Ensure only failure in automated tests is from the pull request for differences due to new output from the new use case

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[Yes]**</br>

After merging PR, develop-ref branch will need to be updated to include the new output generated by the new use case.

- [X] Please complete this pull request review by **11/15/2021**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: metplus/util/met_util.py [hunk 1]
```diff
@@ -1694,7 +1694,7 @@ def getlist(list_str, expand_begin_end_incr=True):
         return []
 
     # FIRST remove surrounding comma, and spaces, form the string.
-    list_str = list_str.strip().strip(',').strip()
+    list_str = list_str.strip(';[] ').strip().strip(',').strip()
```
Comment (JohnHalleyGotway) [inline metplus/util/met_util.py:1697]: Want to make sure this change is intentional. I assume it is, but this is the sort of the small change that could wreak havoc if not intended!

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

Wow, this is a whopper. One PR to resolve 4 issues!

I see that diffs for one of the MET Tool Wrapper tests failed, as expected, since new output was created:
```
Comparing output to truth data
ERROR: File does not exist: /data/truth/IODA2NC/metplus_final.conf
ERROR: File does not exist: /data/truth/IODA2NC/ioda2nc/ioda.NC001007.2020031012.summary.nc
```
I noticed that docs/use_cases/met_tool_wrapper/IODA2NC/README.rst is basically empty. But I do see that all other similar README files are also empty.

I also added a comment about the change in met_util.py. You could update the comment to explain why stripping semicolons and square brackets is done there.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

Wow, this is a whopper. One PR to resolve 4 issues!

I see that diffs for one of the MET Tool Wrapper tests failed, as expected, since new output was created:
```
Comparing output to truth data
ERROR: File does not exist: /data/truth/IODA2NC/metplus_final.conf
ERROR: File does not exist: /data/truth/IODA2NC/ioda2nc/ioda.NC001007.2020031012.summary.nc
```
I noticed that docs/use_cases/met_tool_wrapper/IODA2NC/README.rst is basically empty. But I do see that all other similar README files are also empty.

I also added a comment about the change in met_util.py. You could update the comment to explain why stripping semicolons and square brackets is done there.

File: metplus/util/met_util.py (hunk 1 above)
Comment (georgemccabe) [reply metplus/util/met_util.py:1697]: Yes, this is the change that resolves #1212 by stripping off square braces and semi-colons from the list value.

**Pull request merged**